### ISSUE TITLE
[css-sizing] Adding a 'size' shorthand for 'width'/'height'

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#820]
Many people thruout the years have requested a 'size' property as a shorthand for 'width' and 'height', for the common cases when you want to set them both to the same value.  (Either making the element square, or making it a similar rectangle to the containing block with percentages, or using the same keyword for both.)

This seems mildly useful, but it has a naming problem - the @page rule already has a 'size' property, and it's *not* a shorthand for the 'width' and 'height' properties @page also has. ('size' sets the size of the margin box, while 'width' and 'height' set the content box.)

Other suggestions for names?

**Comment by astearns** [comment]
dimensions (probably too long and difficult to spell)
area (not quite the right meaning)
embiggedness (the real reason I'm adding this comment)

I've added this to the list of mistakes

**Comment by frivoal** [comment]
Also, just to put it out there, how bad would it be if we added this under the `size` name **and** kept `@page`'s `size` stay the way it is, effectively making the `size` shorthand unavailable in `@page`?

On the one hand, this is dirty and there is no precedent for properties meaning completely different things in different context, on the other hand, the usages seem really disconnected, and authors would likely cope just fine.

(and on the third hand, it isn't entirely clear why `size` in `@page` is considered a property rather than a descriptor, but it doesn't really make much of a difference anyway, so ignore my 3 handed monster).

**Comment by Crissov** [comment]
`content-box`

**Comment by frivoal** [comment]
Not `content-box`, as that would be the wrong name when `box-sizing: border-box` is on.
As for `box`, it doesn't seem wrong, but it's awfully generic.

**Comment by inoas** [comment]
I don't see the benefit. I'd rather not pollute the property space with property names that yield little benefit. If you want to set both, then just set both.

p.s.: What would be more interesting is aspect-ratio values for height and width (behaving similar to auto, insofar as that they will scale when one value is set or the other).

**Comment by Crissov** [comment]
Without added features beyond a shorthand, such a new property would be [nonsense anyway](https://github.com/w3c/csswg-drafts/issues/551#issuecomment-253339556).

    <<box>>: [ <<width>> <<height>>? | <<width>> <<box-aspect>> | <<box-aspect>> <<height>> ] || <<box-sizing>>
    <<box-aspect>>: <number> | <integer> '/' <integer> | 'golden' | 'sqrttwo'

**Comment by inoas** [comment]
What would take precedence if width/height and `new-property` are set and values are conflicting?

I'd argue something like this would make sense:
img.some {
}

**Comment by inoas** [comment]
What would take precedence if width/height and `new-property` are set and values are conflicting?

I'd argue something like this would make sense:
```css
.block {
  height: 640px; /* change this through MQ */
  width: 480px;  /* change this through MQ */
}
img.some {
  display: block;
  width: 100%;
  height: aspect(0.5);      /* set's the height to half of the width */
  height: auto(0.5);        /* set's the height to half of the width */
  height: auto-aspect(0.5); /* set's the height to half of the width */
  height: auto-ratio(0.5);  /* set's the height to half of the width */
  height: auto-ratio(golden);  /* set's the height to golden cut */
  height: auto-ratio(calc(9/16));  /* set's the height to 16:9 */
}
```

**Comment by tabatkins** [comment]
This topic is not about a theoretical aspect-ratio property; please move any such discussion to a different issue.  (There might already be one started for that.)

**Comment by inoas** [comment]
@tabatkins so on the constructive side:
What would take precedence if width/height and new-property are set and values are conflicting?
What are the functional benefits?

**Comment by Crissov** [comment]
The issues are connected however, @tabatkins, because it could make sense to name them all `box-*` and combine them in a `box` shorthand property – or only add that. The syntax of an aspect ratio property or function should be discussed separately, though.

    <box>: <box-size> || <box-sizing>
    <box-size>: [ <width> <height>? | <width> <box-aspect> | <box-aspect> <height> | <box-aspect> ] /* = */
    <box-size>: [ <width> [ <height> | <box-aspect> ]? | <box-aspect> <height>? ]

**Comment by inoas** [comment]
@Crissov then the rule would be whatever is set as a property (with the same or higher specificity, obviously) overwrites each other? E.g. would this hold true?

```css
div {
  dimensions: 100px 200px; /* width: 100px, height 200px */
  width: 50px;  /* width: 50px, height: 200px */
  dimensions: 100px; /* width: 100px, height: 200px */

  /* Related / Off-Topic: */
  dimensions: 100px aspect-ratio(calc(9/16)); /* width: 100px, height: 56.26px */
}
```

**Comment by Crissov** [comment]
I'd say that a single length makes the box square -- assuming 1:1 is the default aspect ratio that the shorthand resets to: 
100px * 200px @ 1:2 --> 50px * 200px or 50px * 100px --> 100px * 100px.

It could also keep the current AR, hence 
100px * 200px --> 50px * 200px --> 100px * 400px.

Alas, then `width` before should do, too, i.e. 
100px * 200px @ 1:2 --> 50px * 100px --> 100px * 400px. 

Otherwise, yes `box` and `box-size` would work like other shorthand properties. The WG would decide the most useful and compatible approach.

**Comment by fantasai** [comment]
@inoas Precedence is calculated by the cascade, as `size` would be a shorthand property. See https://www.w3.org/TR/css-cascade/#shorthand-property

@Crissov As Tab said, please take discussion of aspect ratios elsewhere. This issue is literally just a naming conflict issue.

**Comment by fantasai** [comment]
@inoas Oh, as for the benefit: there are many times when one wants to set both width and height, or both min/max-width and min/max-height, to the same thing. It's not commonly desired for length values (since that would always make a square), but for keywords and percentages it is.

**Comment by Crissov** [comment]
@fantasai The detour to aspect ratios was necessary, for me at least, to prefer `box-size` over just `box` as an alternative to `size`. @tabatkins didn’t specify the exact syntax or scope of the shorthand nor link to one, so it’s difficult to suggest an alternative without considering such side issues.

**Comment by fantasai** [comment]
It would be `size: <'height'> <'width'>?` with omitted `<'width'>` taken from `<'height'>`.

**Comment by inoas** [comment]
@fantasai `<'height'> <'width'>?` would be consistent with how padding and margin with 2 values works in terms of the direction (horizontal, vertical), however elsewhere sizes are usually defined width first then height. That's the only thing that would make me worry.

Thanks for pointing to the shorthand block. I just wasn't aware of the nomenclature. That clears the conflicting (that isn't there!).

**Comment by jonathantneal** [comment]
> this is dirty and there is no precedent for properties meaning completely different things in different context

Are there other concerns with @frivoal’s suggestion that the `@page` at-rule could define its own context where `size` uses the existing meaning? To me, this seems worth exploring, as the alternatives seem far from the cowpath.

**Comment by Crissov** [comment]
@inoas The similarity of `box-size` and `box-sizing` would be a feature, not a bug, in my opinion – also possible:

    box-size: <'height'> <'width'>? || <'box-sizing'>?

**Comment by fantasai** [comment]
Note also that `inline-size` and `block-size` use the word `size`, so using any other term would be inconsistent with those.

**Comment by tabatkins** [comment]
Conclusion from today's call: currently tabling this issue.  We can unblock it with:

* implementor interest in using 'size', despite the minor implementation concerns that it brings due to context-difference between normal styles and `@page`
* a new name that meets broad agreement.

**Comment by jonathantneal** [comment]
What does implementor interest look like? Are outsiders (those who don’t work with browser vendors) capable of helping in this area?

If an implementor is interested, how blocking is the later issue. I’m just not clear if this is an either/or list.

**Comment by astearns** [comment]
It's an either/or list. 

If we can come up with a name that isn't 'size' that is really good for the author experience, then it's relatively easy to spec and implement.

Or if there is enough author convenience in a shorthand named 'size' that convinces implementors to take it on, we can progress with that name. My understanding (as a non-implementor) is that this approach is slightly more complicated, and no one has (yet) considered this shorthand worth the effort to implement. Bugs, blog posts and bribes are needed.

**Comment by upsuper** [comment]
I have another concern for serialization. If we add `size` (or whatever it would be called), the following block
```css
div {
  width: 200px;
  height: 100px;
}
```
would suddently be serialized to something like `size: 200px 100px;`.

Code may be broken if they rely on parsing the serialization themselves. Not sure how usual this can be, though.

**Comment by Crissov** [comment]
Naming isn’t really something that implementers, i.e. browser vendors, should have the last say in. Like @astearns says, it’s about the “author experience”. 

> ## [Module interactions](https://drafts.csswg.org/css-sizing/#placement)
> This module extends the `width`, `height`, `min-width`, `min-height`, `max-width`, `max-height`, […] features defined in [CSS21] chapter 10 […]

The respective level-3 module to replace the level-2 reference would be [[CSS-**BOX**]](https://drafts.csswg.org/css-box/) if I’m not mistaken. At the risk of sounding annoying, the `box-` prefix (and `box` shorthand) seems appropriate to me, although it is implied in most properties currently.

(Fun fact: `box-sizing` was put into [[CSS-UI]](https://drafts.csswg.org/css-ui/#propdef-box-sizing) back in the day – i.e. 200*X* –, just because it was thought to be one of the earliest modules to reach REC status that wasn’t feature-frozen already.)

**Comment by frivoal** [comment]
> ( Fun fact: box-sizing was put into [CSS-UI] back in the day – i.e. 200X –, just because it was thought to be one of the earliest modules to reach REC status that wasn’t feature-frozen already.)

While many modules have made a lot of progress since, not many have reached REC, so this may still be true :) `</digression>`

**Comment by cvrebert** [comment]
Another possible name: `diameter`

**Comment by jonathantneal** [comment]
@fantasai, @tabatkins, was `box-size` presented as a name? I don’t see where it was disqualified. Is there an the issue? It also seems to address [these concerns](https://github.com/w3c/csswg-drafts/issues/820#issuecomment-281738756) (repeated below):

> Note also that `inline-size` and `block-size` use the word `size`, so using any other term would be inconsistent with those.

**Comment by Crissov** [comment]
@jonathantneal Yes, in this very thread.

**Comment by tomhodgins** [comment]
I've toyed around with this idea too - using a shorthand for both `width` and `height` called `size` that sets a value for both, kind of like how JavaScript can express either of these:

```javascript
var width = 100
var height = 100
```

and also:

```javascript
var width = height = 100
```

However, I don't think what I was really searching for was actually a `size` property, but a syntax that would let me 'chain' properties that share the same value. While I think the most-often requested use case for property chaining would be the `size` property, what if chaining was the new feature instead of just one new 'chained' property? Imagine this:

```css
div {
  width: 100px;
  height: 100px;
}
```

Can already become something like this thanks to CSS variables:

```css
div {
  --size: 100px;
  width: var(--size);
  height: var(--size);
}
```

That functions the same way we want, but you have to write more CSS to make them share the same variable here than just setting them to the same number separately. But what about something like:

```css
div {
  width, height: 100px;
}
```

I've just use a comma `,` here for a delimiter (maybe there's a better way) but this might be a more flexible solution for expressing this sort of thing than inventing new combined properties one at a time like `width` + `height` = `size`. What if CSS authors had the expressive ability to chain any two or more properties to set the same value?

**Comment by jonathantneal** [comment]
@Crissov, ah, sorry, I meant in a CSSWG meeting. See https://twitter.com/csswg/status/897994427859214336

**Comment by tomhodgins** [comment]
I was inspired to build a little demo of this using `--size`:

```html
<div>Demo</div>

<style>
    div {
    background: lime;
    --size: 50vmin;
  }
</style>

<script>
  function sizer() {

    var tag = document.querySelectorAll('body *')

    tag.forEach(function(tag) {

      var currentStyle = window.getComputedStyle(tag)
      var sizeProperty = currentStyle.getPropertyValue('--size')

      tag.style.width = sizeProperty
      tag.style.height = tag.offsetWidth + 'px'

    })

  }

  window.addEventListener('load', sizer)
  window.addEventListener('resize', sizer)
  window.addEventListener('input', sizer)
  window.addEventListener('click', sizer)
</script>
```

Demo: https://codepen.io/tomhodgins/pen/WEdmMM

**Comment by battaglr** [comment]
@tabatkins, is this still being considered? 

If we added `size` —or whatever it would be called— shouldn't we also consider adding `min-size` and `max-size`?

**Comment by battaglr** [comment]
Related: [Logical Height and Logical Width: the `block-size` and `inline-size` properties](https://drafts.csswg.org/css-logical/#dimension-properties).

**Comment by hacknug** [comment]
Since we're dealing with rectangles, maybe we could go with `side`?

**Comment by fantasai** [comment]
Suggestion to resolve the @page { size: ... } naming conflict by internally renaming the page descriptor to `page-size` at parse-time or thereabouts, leaving `size` free for use everywhere else as a normal property.

**Comment by heycam** [comment]
Suggestion seems fine to me.  To clarify, `CSSPageRule.style.getPropertyValue("size")` would return an empty string, and you'd need to pass the real descriptor name `"page-size"` to get its value?

**Comment by tabatkins** [comment]
> To clarify, CSSPageRule.style.getPropertyValue("size") would return an empty string, and you'd need to pass the real descriptor name "page-size" to get its value?

Yes. My intention (to be clarified in CSSOM working with @emilio or @zcorpan) is that anything which parses a whole declaration or larger will allow "size" and translate it, but anything which just takes a declaration name directly won't know about "size".

**Comment by Marat-Tanalin** [comment]
Hopefully both single-value (`size: 10px`) and _two-values_ (`size: 2em 3em`) syntaxes are going to be supported.

**Comment by tabatkins** [comment]
Yes, absolutely.

**Comment by bernhardf-ro** [comment]
I am with RealObjects. If we may weigh in on the actual usage of CSS Paged Media: Our product PDFreactor, a HTML-to-PDF converter, has been commercially available since 2006 and has supported JavaScript since 2012. Other commercial paged media user agents include Prince, supporting JavaScript since 2011, and Antenna House Formatter.

The introduction of the '`page-size`' property and the new behavior of '`size`' would be an API change for the CSSOM and derived data structures. For functionalities that have been in production use for years, we'd appreciate if significant changes could be avoided.

We'd like to suggest having another look at the approach where '`size`' is a shorthand everywhere and the `width` and `height` properties apply to pages, thus eliminating the need for the '`page-size`' property. We realize that `width` and `height` would be applied to the margin box for pages. However, we don't think that this would create significant confusion. Most importantly, this approach would not cause any breaking changes for authors at all. Please note that we even encountered customers that found it counter-intuitive that width and height cannot be set on pages.

**Comment by Marat-Tanalin** [comment]
@bernhardf-ro

> this approach would not cause any breaking changes for authors at all.

Afaict, the change is _not_ breaking since `size` inside `@page` will _continue_ to work for the purpose of backward compatibility, effectively being an _alias_ for the new `page-size`:

> Also defining that size in the page context parses into page-size

Still, I agree that renaming `size` to `page-size` is probably not necessary, given that `size` meaning inside `@page` rules is similar to its new meaning for regular elements, and especially given that `size` and `page-size` are going to be interchangeable inside `@page` _anyway_, so adding `page-size` probably just bloats the CSS keyword list.

**Comment by faceless2** [comment]
According to https://www.w3.org/TR/css-page-3/#fixed-sizing, `width` and `height` do apply to paged media. If they are specified, they set the size of the page "content" box, and `size` is ignored.

Just for completeness, because I haven't seen it suggested yet:

`size` in the `@page` context is effectively setting the margin box on the page. I don't know how often the `width` and `height` properties are used instead to specify page sizes, but as this option is new in CSS3 I'd expect it's a lot less common than `size`.

So another option would be to introduce `margin-box` as a new values for `box-sizing`, and use that as the default in the page context. If you allow for a few special values (like "A4" as an alias for "210mm 297mm"), then size becomes the shorthand for width/height without any need to rename anything. The cost is if anyone is using `width` and `height` to set the page size, they would need to add a new `box-sizing: context-box` declaration to their `@page` rule.

**Comment by bernhardf-ro** [comment]
Documents accessing the `size` property of an `@page` rule in the CSSOM via JS will stop working. So at least technically the change is breaking. We do not have enough data for a detailed estimate on the magnitude of the actual negative effects, but there definitively will be more than none. (We also offer a Paged Media based editing framework, which allows more direct access to the CSSOM for more complex integrations. Those customers would also be affected. So it is not only JavaScript.)

Thanks for pointing out that in the Paged Media specification `width` and `height` already affect the page size, specifically the page area (page content box). However, there is practically no support for this in user agents while `size` is supported and essential in all Paged Media based ones. Also we are not sure about the use cases for basing the page size on its area/content size. So, in our opinion, changing `width` and `height` looks preferable to changing `size`.

Also the behavior of `width` and `height` already depends on the context in multiple cases, like inline or table related boxes. Having them affect the margin size, instead of the one from `box-sizing`, when used in `@page` is not a complex or confusing special case. (It may even be what authors expect, but that is speculation.)

We also agree that paper sizes like `A4` can be handled at the shorthand level and think that would be an elegant solution, also allowing to style any box to have one of the standardized sizes.

Overall the approach of having the shorthand in all cases, including `@page`, may have more positive effects and opportunities than negative ones.

**Comment by tabatkins** [comment]
> Documents accessing the size property of an @page rule in the CSSOM via JS will stop working. So at least technically the change is breaking. 

Yes, `pagerule.getPropertyValue('size')` would stop working; this is indeed a breaking API change.

> We also agree that paper sizes like A4 can be handled at the shorthand level and think that would be an elegant solution, also allowing to style any box to have one of the standardized sizes.

A problem here is that shorthands have to re-synthesize themselves from the longhands, but the keywords do *not* map cleanly to integer `px` values (and lengths in all UAs are stored as some fixed fraction of a `px`), which means that if you, say, set `size: A4;`, you'll read back two decimal-heavy `px` lengths unless we specify a "close enough" rounding that must return the keyword.

(That's not impossible, but it is new.)

> [suggestion to switch @page/width and height to setting the page margin box and making them shorthands of size]

Hm, that's interesting. If you're right that the usage of width/height on @page is lower (and thus will cause less breakage) than the JS querying of size, then that's a reasonable alternative. Getting some feedback from other printing software would be useful here.

**Comment by bernhardf-ro** [comment]
> Getting some feedback from other printing software would be useful here.

We also hope for more parties commenting on this. Maybe adding `[css-page-3]` to the title would draw the necessary attention.

If it is any help, we found that of 4 user agents that support `size` (Chrome, PDFreactor, Prince, AH Formatter), 3 do not support `width` and `height` at all and the remaining one, AH Formatter, does not resize the page as per the specification.
So, while there is one implementation that does not ignore `width` and `height`, its behavior deviates far enough from the specification that making it compliant to the current specification would be as much of a change as adapting it to our proposal.


> [what exactly retrieving `size` should return]

Returning 2 lengths, even after setting an identifier, looks like the best solution to us. Most script authors that have to retrieve the `size` would benefit from a consistent format and most likely prefer always getting lengths. 
For example, `size: A4`, just like `size: 210mm 297mm`, should result in `width: 210mm; height: 297mm` and, in turn, in `size: 210mm 297mm`.

On the other hand, to keep the identifier if (and only if) it has been set, would be possible, but require passing it to `width` and `height`, e.g. `width: A4; height: A4`, which we consider excessive. It would create a lot of additional work for the sake of symmetry.

**Comment by Crissov** [comment]
Paper size approximation 
-----------------------------

Regarding standard page sizes being stored with the canonical unit `px` (in browsers at least, hopefully not in CSS-to-PDF software): since `24px` = `0.25in`, US sizes, which are all based on full, half or quarter inches, do not have the problem, but `calc(96px/25.4)` and `calc(25.4mm/96)` yield indeed awfully odd numbers, roughly `3.78px` and `0.265mm`. A4, i. e. 297 mm × 210 mm, would become c. `1122.52px` × `793.79px`, which, if first rounded to the nearest CSS pixel, converts back to `297.12mm` × `210.08mm`. 

Many PDFs exhibit a similar issue, wherein ISO page dimensions have been converted to an integer number of DTP points, resulting in 297.04 mm × 209.9 mm for instance.

 However, ISO 216 specifies tolerances for cut paper sizes that are well beyond such rounding effects: ±1 mm for the sizes that have keywords in CSS if I remember correctly. Also, the sizes are iteratively rounded (floored?) from theoretical √2-based values to an integer number of millimeters anyway. The theoretical size of A4 would be about 297.302mm × 210.224mm or ca. `1123.66px` × `794.55px`. 

In other words, if width and length are within 1mm of an ISO size after having been converted from a `px`, itʼs pretty safe to assume that the size originally specified was indeed the respective standard format. 

New `size` keywords 
----------

This reminded that there are other common sizes that may deserve keywords. (See #328 for paper sizes.) Iʼm not sure theyʼre worth the cost, because their actual values are often at least as familiar as their names and abbreviations, but Iʼm thinking of `vga` = `640px 480px` and the like. Besides [`<page-size>`](https://drafts.csswg.org/css-page-3/#typedef-page-size-page-size), `size` should then also learn orientation modifiers `landscape` and `portrait` from `@page`.

PS: These days, I’d also expect standard (IAB) [ad sizes](https://support.google.com/admanager/answer/1100453). 

<details>

- `leaderboard` = `728px 90px`
- `large-leaderboard` = `970px 90px`
- `large-rectangle`  = `336px 280px`
- `medium-rectangle` = `300px 250px`
- `small-rectangle`  = `180px 150px`
- `vertical-rectangle` = `240px 400px` 
- `tablet-landscape`, full-screen = `1024px 768px`
- `tablet-portrait` = `768px 1024px`
- `banner` = `468px 60px` 39:5
- `half-banner` = `234px 60px`
- `vertical-banner` = `120px 240px`
- `mobile-landscape` = `480px 320px`
- `mobile-portrait` = `320px 480px`
- `mobile-leaderboard` = `320px 50px`
- `mobile-leaderboard-double` = `320px 100px`
- `mobile-banner` = `300px 50px`
- `mobile-banner-double` = `300px 100px`
- `half-page` = `300px 600px`
- `wide-skyscraper` = `160px 600px`
- `skyscraper` = `120px 600px`
- `square` = `250px 250px`
- `small-square` = `200px 200px`
- `square-button` = `125px 125px`
- `button` = `120px 60px`
- `small-button` = `88px 31px`
- `panorama` = `980px 120px`
- `top-banner` = `930px 180px`
- `netboard` = `580px 400px`
- `triple-widescreen` = `250px 360px` 
</details>

**Comment by XiBOR** [comment]
Wouldn't it be more convenient to make reductions for all properties at once?
For example:
```
size: 10px; // width: 10px, height: 10px || min-width: 10, max-width: 10, min-height: 10, max-height: 10
size: 10px 20px; // width: 10px, height: 20px || min-width: 10, max-width: 10, min-height: 20, max-height: 20
size: 10px 20px 30px 40px; // min-width: 10, max-width: 20, min-height: 30, max-height: 40
```

**Comment by Crissov** [comment]
`size: 10px` could also be a shorthand for roughly `size: 7.07px 7.07px` with an aspect ratio of 1:1, i. e. specify the diagonal of diameter.

**Comment by Loirooriol** [comment]
@Crissov You can use `calc(10px * cos(atan(1/1)))` and `calc(10px * sin(atan(1/1)))` for that.
Overcomplicating `size` with diagonals would be very confusing and unexpected.

**Comment by Crissov** [comment]
Diagonals would only make sense if one could also set the aspect ratio within `size`:

```
size: [<'width'> <'height'>] 
    | [<length> || <'aspect-ratio'>] 
```

**Comment by tabatkins** [comment]
I'm downloading AH Formatter right now (well, I'm requesting a trial download, we'll see), so I can see what they actually do with 'width'/'height' on pages, per @bernhardf-ro's [comment](https://github.com/w3c/csswg-drafts/issues/820#issuecomment-545463628) that they're the only implementor that has those descriptors at all, and they do something different than the spec anyway.

Changing the definition of @page/width and @page/height seems like the most viable way forward here, since no one's come up with a reasonable alternate name for 'size'.

**Comment by zjopy** [comment]
I don't think it was mentioned yet, but the css-writing-mode module uses the word "size" in `block-size` and `inline-size` as a length in a single dimension. This was a really unfortunate choice as "size" refers to "a thing's overall dimensions or magnitude" [source](https://www.lexico.com/en/definition/size), which for 2-dimensional boxes, would mean 2 dimensions. It definitely should have been named "block/inline-length" or similar, but I guess it's too late now.

This however means that, while entirely reasonable on its own, using `size` as a shorthand for both `width` and `height` would conflict with the one-dimensional notion used in the css-writing-mode module.

I guess nobody cares though. There are already tons of conflicting naming choices in CSS, so yet another one won't be that bad.

**Comment by jsnkuhn** [comment]
Would there be a need for `min-size` and `max-size` properties as shorthands for `min-width`/`min-height` and `max-width`/`max-height`? Or would the use of `size:  clamp()` or `size:  clamp() clamp()` be enough to cover all of that?

**Comment by SebastianZ** [comment]
Using `clamp()`, `min()`, and `max()` should be sufficient to cover all use cases of `min-width`, `min-height`, `max-width`, and `max-height`.

Sebastian

**Comment by Loirooriol** [comment]
Note that `width: 100px; max-width: 0%` and `width: min(100px, 0%)` have different intrinsic contributions when the containing block is intrinsically sized. https://drafts.csswg.org/css-sizing/#cyclic-percentage-contribution

**Comment by jonathantneal** [comment]
I see this issue now has the tag **Needs Design / Proposal**. What might that _design_ or _proposal_ look like? Would it look like a new issue with the proposed changes as they might appear in a bikeshed document? Or as a PR?

Perhaps some of us can help with this. 😄

**Comment by MurakamiShinyu** [comment]
I am not opposed to changing the definition of @page/width and @page/height. However, we still need a way to specify the width and height of the page-area. For example, what about the following new syntax?  (`@page-area` for the page area definition):

```css
@page {
  size: A5;
  margin: auto;
  @page-area {
    width: 32rem;
    height: 29rlh;
  }
}
```

BTW, our implementation, [Vivliostyle](https://vivliostyle.org/), has supported @page/width and @page/height in the same way as Antenna House Formatter since 2015.

See our sample stylesheet using width and height on `@page` : 
https://github.com/vivliostyle/vivliostyle_doc/blob/gh-pages/samples/gon/scss/base/_format.scss#L23-L32
(View the sample with Vivliostyle Viewer: https://vivliostyle.org/viewer/#src=https://vivliostyle.github.io/vivliostyle_doc/samples/gon/index.html )

**Comment by fantasai** [comment]
I chatted with @emilio about this back in December. IIRC, it seems it's possible to just have `size` in an `@page` context just be handled specially. It would be a little bit confusing that outside of `@page` it's a shorthand for width+height, but within `@page` it's a special property, but I think it might be the best option. I'd rather not define `@page` `width`/`height` in a way that's incompatible with the CSS-to-PDF renderers that have implemented it so far. And we've failed to come up with a better term than `size` for the width+height shorthand. If that seems practical to other implementers, maybe we can move forward with that.

**Comment by fantasai** [comment]
@MurakamiShinyu @faceless2 What do you think?

**Comment by MurakamiShinyu** [comment]
@fantasai I agree that would be the best option, to keep the current definitions in `@page` context while allowing the new `size` shorthand in other context.

**Comment by faceless2** [comment]
It's not ideal, but I agree it's probably the simplest option.

**Comment by Crissov** [comment]
Isn’t `size` inside `@page` more like a descriptor than like a property anyway?

**Comment by bernhardf-ro** [comment]
We discussed this internally again and still think that just making `size` a shorthand and having page margin box size be affected by `width` and `height` would be the ideal solution.
Making the property a shorthand depending on its context in the CSS-OM would add complexity that may well be disproportionate to the gain.
However, we don't see any issues on the implementation side that would prevent us from implementing it if it is decided on.

**Comment by jsnkuhn** [comment]
CSS Grid has a `grid` shorthand property that allows for defining `display: grid;` related properties in one line could there similarly be a `block` property related to `display: block;` that covers width/height, alignment, margin/padding?

https://developer.mozilla.org/en-US/docs/Web/CSS/grid

**Comment by ggedde** [comment]
Maybe I am missing something here, but can't @page {size} just get ignored if the values are in (px, em, rem, %, etc) and vise versa for block {size} if the values are in (A5, A4, Letter, mm, in, etc)

**Comment by cdoublev** [comment]
> It seems it's possible to just have `size` in an `@page` context just be handled specially. It would be a little bit confusing that outside of `@page` it's a shorthand for `width`+`height`, but within `@page` it's a special property, but I think it might be the best option.

Specs do not seem to preclude handling it specially. [CSS Syntax](https://drafts.csswg.org/css-syntax-3/#ref-for-typedef-declaration-list%E2%91%A3) requires that a CSS rule taking `<declaration-list>` define the properties and descriptors it accepts: `@page` does not accept the `size` property but accepts a `size` descriptor.

*From this perspective*, I do not see where CSSOM prevents a context-specific handling (even if `CSSPageRule.style`, ie. `CSSStyleDeclaration`, is defined with all properties). `font-style` already has a slightly different syntax depending on if it is declared in a style rule or `@font-face`, and `CSSFontFaceRule.style.setPropertyValue('font-style')` is supposed to be supported (`style` and `setPropertyValue()` are misleading).

Fwiw, what confuses me is not that a property grammar could be different depending on the context, but it is this differentiation between property and descriptor, which I cannot find in any other language, and to support `CSSPageRule.style.setPropertyValue('size')` at the same time (`style` is misleading).

**Comment by tabatkins** [comment]
> Fwiw, what confuses me is not that a property grammar could be different depending on the context, but it is this differentiation between property and descriptor, which I cannot find in any other language,

It's largely a legacy separation. For whatever reason, the original editors thought it was valuable to draw a distinction between the two, and we've continued with it since then.

**Comment by andruud** [comment]
What is blocking us here now?

> > Documents accessing the size property of an @page rule in the CSSOM via JS will stop working. So at least technically the change is breaking.
> 
> Yes, `pagerule.getPropertyValue('size')` would stop working; this is indeed a breaking API change.

Is it the above? If yes, can't we "just" do the following?

> IIRC, it seems it's possible to just have `size` in an `@page` context just be handled specially. It would be a little bit confusing that outside of `@page` it's a shorthand for width+height, but within `@page` it's a special property, but I think it might be the best option.

Making a distinction between property/descriptor seems acceptable, especially with [recent developments](https://github.com/w3c/csswg-drafts/commit/2e8c5020cffb6e0e3bd32d756ec08ed1b1968d02) in CSSOM.

**Comment by benface** [comment]
> Note also that `inline-size` and `block-size` use the word `size`, so using any other term would be inconsistent with those.

Just a thought: if `size` is introduced as a shorthand, the `inline-size` and `block-size` properties would be slightly inconsistent with other logical properties where the `inline` and `block` keywords _follow_ the shorthand name, rather than precede it (e.g. `padding-inline`, `margin-block`, etc.). So I think it would make sense to also introduce `size-inline` and `size-block` aliases as well.

**Comment by phaux** [comment]
@benface `gap` also has its more specific versions the other way around: `row-gap` and `column-gap`

**Comment by stof** [comment]
should `size` be a shorthand for `<width> <height>` or for `<inline-size> <block-size>` ?

**Comment by mstensho** [comment]
This means that this version of `size` can never be part of https://drafts.csswg.org/css-page-3/#page-property-list , which is a bit weird, given that its longhands `width` and `height` are on that list.

**Comment by hacknug** [comment]
> should `size` be a shorthand for `<width> <height>` or for `<inline-size> <block-size>` ?

Only for `<width> <height>` atm. From the meeting logs:

> <TabAtkins> ntim: woudl longhand be physical or logical
<TabAtkins> fantasai: i think would have to be physical because that's how we generally handle shorthand properties with both directions
<TabAtkins> fantasai: we map to the physical directions by default
<TabAtkins> florian: would be good to have a switch but we dont' have yet
<TabAtkins> ntim: possible confusion with inline-size/block-size, versus width/height
<TabAtkins> ntim: but otherwise fine with this

**Comment by JoshTumath** [comment]
> should `size` be a shorthand for `<width> <height>` or for `<inline-size> <block-size>` ?

That was asked in the call as well. It would need to be width and height, to follow the president set by other shorthands like `margin` and `padding`. There is a separate issue, #1282, to solve that problem.

> This means that this version of `size` can never be part of https://drafts.csswg.org/css-page-3/#page-property-list , which is a bit weird, given that its longhands `width` and `height` are on that list.

@emilio has filed #11925 to make a new property to replace `size` for `@page`.

**Comment by astearns** [comment]
> > This means that this version of `size` can never be part of https://drafts.csswg.org/css-page-3/#page-property-list , which is a bit weird, given that its longhands `width` and `height` are on that list.
> 
> [@emilio](https://github.com/emilio) has filed [#11925](https://github.com/w3c/csswg-drafts/issues/11925) to make a new property to replace `size` for `@page`.

If we do alias `size` in `@page` as that issue suggests, we would still not be able to use the property version of `size` as a descriptor in `@page`

**Comment by KevinDoughty** [comment]
Please for the love of all things holy allow animating size with negative width and height components

**Comment by benface** [comment]
> Please for the love of all things holy allow animating size with negative width and height components

Huh?

**Comment by KevinDoughty** [comment]
> Huh?

Relative Animation. It’s like FLIP except additive and would work for all properties not just transforms, using the easing function to blend animations when interrupting and retargeting. W3C has not shown interest, but negative width and height would allow manually constructing these animations with Web-Animations, which otherwise are only possible for translation and opacity

https://openradar.me/12085417
https://lists.w3.org/Archives/Public/public-fx/2013AprJun/0211.html
https://github.com/w3c/csswg-drafts/issues/10646
https://github.com/w3c/csswg-drafts/issues/10927
https://github.com/w3c/csswg-drafts/issues/11111
https://github.com/w3c/csswg-drafts/issues/11087
https://github.com/w3c/csswg-drafts/issues/10681

**Comment by benface** [comment]
That seems completely unrelated to the current issue though?
